File: DynamicMenu/DynamicMenu/CustomRibbon.cs
using System;
using System.Drawing;
using System.IO;
using System.Resources;
using System.Reflection;
using System.Runtime.InteropServices;
using Application = NetOffice.ExcelApi.Application;
using ExcelDna.Integration.CustomUI;
using NetOffice.VBIDEApi;

namespace MyExcelAddin
{
    [ComVisible(true)]
    public class CustomRibbon : ExcelRibbon
    {
        private Application _excel;
        private IRibbonUI _thisRibbon;

        public override string GetCustomUI(string ribbonId)
        {
            _excel = new Application(null, ExcelDna.Integration.ExcelDnaUtil.Application);
            string ribbonXml = GetCustomRibbonXML();
            return ribbonXml;
        }

        private string GetCustomRibbonXML()
        {
            string ribbonXml;
            var thisAssembly = typeof(CustomRibbon).Assembly;
            var resourceName = typeof(CustomRibbon).Namespace + ".CustomRibbon.xml";

            using (Stream stream = thisAssembly.GetManifestResourceStream(resourceName))
            using (StreamReader reader = new StreamReader(stream))
            {
                ribbonXml = reader.ReadToEnd();
            }

            if (ribbonXml == null)
            {
                throw new MissingManifestResourceException(resourceName);
            }
            return ribbonXml;
        }

        public void OnLoad(IRibbonUI ribbon)
        {
            if (ribbon == null)
            {
                throw new ArgumentNullException(nameof(ribbon));
            }

            _thisRibbon = ribbon;

            _excel.WorkbookActivateEvent += OnInvalidateRibbon;
            _excel.WorkbookDeactivateEvent += OnInvalidateRibbon;
            _excel.SheetActivateEvent += OnInvalidateRibbon;
            _excel.SheetDeactivateEvent += OnInvalidateRibbon;

            if (_excel.ActiveWorkbook == null)
            {
                _excel.Workbooks.Add();
            }
        }

        private void OnInvalidateRibbon(object obj)
        {
            _thisRibbon.Invalidate();
        }

        public void OnPressMe(IRibbonControl control)
        {
            using (var controller = new ExcelController(_excel, _thisRibbon))
            {
                controller.PressMe();
            }
        }

//        public Bitmap GetImage(IRibbonControl control)
//        {
//            switch (control.Id)
//            {
//                case "GetCrackerTemplateButton":
//                    return new Bitmap(Properties.Resources.cracker);
//                case "GetGenericTemplateButton":
//                    return new Bitmap(Properties.Resources.generic);
//                case "GetAllAnalysisButton":
//                    return new Bitmap(Properties.Resources.download);
//                case "GetPreviousAnalysisButton":
//                    return new Bitmap(Properties.Resources.copy);
//                case "RunRollupButton":
//                    return new Bitmap(Properties.Resources.rollup);
//                default:
//                    return null;
//            }
//        }
//        public void OnGetRugbyImage
//        //<button id="CustomImage" label="RugbyImage"  getImage="OnGetRugbyImage" size="large" onAction="OnPressMe" />
    }
}

File: HelloWorld_VB.Net/MyExcelAddIn/CustomRibbon.vb
Imports System
Imports System.IO
Imports System.Resources
Imports System.Reflection
Imports System.Runtime.InteropServices
Imports Application = NetOffice.ExcelApi.Application
Imports ExcelDna.Integration.CustomUI

<ComVisible(True)> _ 
Public Class CustomRibbon : Inherits ExcelRibbon
    
    private  _excel As Application
    private  _thisRibbon As IRibbonUI

    Overrides Function GetCustomUI(ribbonId As String) As String
        _excel = new Application(Nothing, ExcelDna.Integration.ExcelDnaUtil.Application)
        Dim ribbonXml As String = GetCustomRibbonXML()
        return ribbonXml        
    End Function

    Private Function GetCustomRibbonXML() As String
        Dim ribbonXml As String = Nothing
        Dim thisAssembly As Assembly = GetType(CustomRibbon).Assembly
        Dim resourceName As String = GetType(CustomRibbon).Namespace + ".CustomRibbon.xml"

        Using stream as Stream = thisAssembly.GetManifestResourceStream(resourceName)
        Using reader As StreamReader = new StreamReader(stream)
            ribbonXml = reader.ReadToEnd()                
        End Using
        End Using

        If (ribbonXml Is Nothing) Then
            throw New MissingManifestResourceException(resourceName)
        End If
        Return ribbonXml
    End Function

    Public Sub OnLoad(ribbon As IRibbonUI)
        _thisRibbon = ribbon
        If (_excel.ActiveWorkbook Is Nothing) Then
            _excel.Workbooks.Add()
        End If
    End Sub

    Public Sub OnPressMe(control As IRibbonControl)
            Using controller As ExcelController = New ExcelController(_excel, _thisRibbon)
                controller.PressMe()            
            End Using
    End Sub

End Class
